pkg/distributor/distributor.go

```diff
@@ -300,6 +305,25 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	}
 }
 
+func (d *Distributor) truncateLines(vContext validationContext, stream logproto.Stream) []logproto.Entry {
+	if !vContext.maxLineSizeShouldTruncate {
+		return stream.Entries
+	}
+
+	n := 0
+	for _, e := range stream.Entries {
+		if maxSize := vContext.maxLineSize; maxSize != 0 && len(e.Line) > maxSize {
+			truncated := float64(len(e.Line) - maxSize)
```

Comment:
From your PR comment:
> This change also doesn't include appending ... truncated 12345 bytes because it has the potential to introduce strange behavior around max_line_length

What do you mean by strange behaviour? If we used `...` as the suffix to indicate truncation, could we not just strip an additional 3 characters off the line?

I think truncating a line without any indication is going to cause confusion to users; I think we should try solve this.

This could even be configurable IMHO

Reply:
I'm worried about the edge-case where the `max_line_size` is `3` or less. In this case, if a log of length 5 came in and we replaces the last 3 non-truncated characters with `...` we'd end up with a log of length `3`. In that case, we'd would either replace the whole log line or it would still above `max_line_size` and we'd drop it line anyway. In either case, it's unhelpful or confusing what's happening to your logs.

That said, if a line is truncated and then dropped, there are metrics for that and maybe that's the right signal to the operator that the max_line_size is too small?

Another option would be to validate that `max_line_size` is at least 4 which would guarantee that  we could replace the last three bytes with `...` without strange behavior

Reply:
I think the approach I'll take is to just ignore truncation if the `max_line_size` is too small and emit a metric about it.

Reply:
I don't think we need to emit a metric for that. It's just a corner case (who would set a line size to under 4 bytes?) I think it can just be covered in a unit test.

Reply:
I agree that `...` is less than ideal. If we had a mechanism to add _unindexed labels_, that'd be great. Maybe in the future...
As for the truncation indicator, this isn't something I feel should block the merging of this nice work, but I do feel that we should provide _some_ indication to users that a line was truncated. Maybe a simple flag for the suffix (defaulting to an empty string) could be the way to split the difference here; that way we put the control in users' hands from the get-go.

Reply:
I've pushed a new commit that has the configurable indicator because it seems like a reasonable compromise.